skops/card/_model_card.py

```diff
@@ -1,61 +1,195 @@
-    table = "| Hyperparameters | Value |\n| :-- | :-- |\n"
-    for hyperparameter, value in hyperparameter_dict.items():
-        table += f"| {hyperparameter} | {value} |\n"
-    return table
 
+    def __init__(self, model, model_diagram=True):
+        self.model = model
+        self.hyperparameter_table = self._extract_estimator_config()
+        # the spaces in the pipeline breaks markdown, so we replace them
+        if model_diagram is True:
+            self.model_plot = re.sub(r"\n\s+", "", str(estimator_html_repr(model)))
+        else:
+            self.model_plot = None
+        self.template_sections = {}
+        self._figure_paths = {}
 
-def create_model_card(
-    model,
-    card_data,
-    **card_kwargs,
-):
-    """Creates a model card for the model and saves it to the target directory.
+    def add(self, **kwargs):
+        """Takes values to fill model card template.
+        Parameters:
+        ----------
+        **kwargs : dict
+            Parameters to be set for the model card. These parameters
+            need to be sections of the underlying `jinja` template used.
+        Returns:
+        --------
+        self : object
+            Card object.
+        """
+        for section, value in kwargs.items():
+            self.template_sections[section] = value
+        return self
 
-    Parameters:
-    ----------
-    model: estimator
-        scikit-learn compatible estimator.
-    card_data: CardData
-        CardData object.
-    card_kwargs:
-        Card kwargs are information you can pass to fill in the sections of the
-        card template, e.g. model_description, citation_bibtex, get_started_code.
-    """
-    ROOT = skops.__path__
-    model_plot = re.sub(r"\n\s+", "", str(estimator_html_repr(model)))
-    hyperparameter_table = _extract_estimator_config(model)
-    card_data.library_name = "sklearn"
-    template_path = card_kwargs.get("template_path")
-    if template_path is None:
-        template_path = os.path.join(f"{ROOT[0]}", "card", "default_template.md")
-    card_kwargs["template_path"] = template_path
-    card = ModelCard.from_template(
-        card_data=card_data,
-        hyperparameter_table=hyperparameter_table,
-        model_plot=model_plot,
-        **card_kwargs,
-    )
-
-    return card
+    def add_plot(self, **kwargs):
+        """Add plots to the model card.
+
+        Parameters:
+        ----------
+        **kwargs : dict
+            The arguments should be of the form `name=plot_path`, where `name`
+            is the name of the plot and `plot_path` is the path to the plot,
+            relative to the root of the project. The plots should have already
+            been saved under the project's folder.
+        Returns:
+        --------
+        self : object
+            Card object.
+        """
+        for plot_name, plot_path in kwargs.items():
+            self._figure_paths[plot_name] = plot_path
+        return self
+
+    def save(self, path):
```

Comment:
I have a few issues with this function, please let me know if you agree:

1. It mutates `self.template_sections`

As a user, I would be surprised to find that the `save` method actually mutates the object itself. At the moment, that might not cause a bug, but this type of hidden mutation has the tendency to come back to bite you. Therefore, I would avoid it except if there is no good reason.

To achieve this, simply declare a copy at the beginning of the method, such as `template_sections = self.template_sections.copy()`. After that, each time you use `self.template_sections` in this method, use `template_sections` instead.

It would be even safer to use `copy.deepcopy` but here we're creating a whole duplicate, I'm not sure if that can be too expensive in some circumstances (doesn't seem so but I'm not sure).

If you agree with this change, here is a unit test that would fail without the change but passes with the change:

```python
import copy
...
def test_template_sections_not_mutated_by_save(destination_path, model_card):
    # TODO
    template_sections_before = copy.deepcopy(model_card.template_sections)
    model_card.save(Path(destination_path) / "README.md")
    template_sections_after = copy.deepcopy(model_card.template_sections)
    assert template_sections_before == template_sections_after
```

2. Remove `if self._figure_paths:` branch

I saw that there is some code duplication that we can avoid. It is not strictly necessary to check `if self._figure_paths` and then do an `else`. The reason it can be avoided is because the `for plot in self._figure_paths` already is a no-op if `self._figure_paths` is empty, so the outcome is the same. Thus this whole code:

```python
        if self._figure_paths:
            with tempfile.TemporaryDirectory() as tmpdirname:
                shutil.copyfile(
                    self.template_sections["template_path"],
                    f"{tmpdirname}/temporary_template.md",
                )
                #  create a temporary template with the additional plots
                self.template_sections[
                    "template_path"
                ] = f"{tmpdirname}/temporary_template.md"
                # add plots at the end of the template
                with open(self.template_sections["template_path"], "a") as template:
                    for plot in self._figure_paths:
                        template.write(
                            f"\n\n{plot}\n"
                            + f"![{plot}]({self._figure_paths[plot]})\n\n"
                        )
                card = ModelCard.from_template(
                    card_data=card_data,
                    hyperparameter_table=self.hyperparameter_table,
                    model_plot=self.model_plot,
                    **self.template_sections,
                )

        else:
            card = ModelCard.from_template(
                card_data=card_data,
                hyperparameter_table=self.hyperparameter_table,
                model_plot=self.model_plot,
                **self.template_sections,
            )
```

can be simplified to:

```python
        with tempfile.TemporaryDirectory() as tmpdirname:
            shutil.copyfile(
                template_sections["template_path"],
                f"{tmpdirname}/temporary_template.md",
            )
            #  create a temporary template with the additional plots
            template_sections["template_path"] = f"{tmpdirname}/temporary_template.md"
            # add plots at the end of the template
            with open(template_sections["template_path"], "a") as template:
                for plot in self._figure_paths:
                    template.write(
                        f"\n\n{plot}\n"
                        + f"![{plot}]({self._figure_paths[plot]})\n\n"
                    )

            card = ModelCard.from_template(
                card_data=card_data,
                hyperparameter_table=self.hyperparameter_table,
                model_plot=self.model_plot,
                **template_sections,
            )
```

There is, however, a small disadvantage to the proposed change: It will result in the temporary copy of the template file to be made even if there is no plot. IMHO this is a small price to pay for the code simplification but feel free to disagree.

3. Breaking down the method

This should probably be done in another PR but I wanted to write it down or else I'll forget:

The `save` method does too many different things at the same time and should be broken down. E.g. there could be one sub-method to get the `card_data` and another one to get the `card`.